### PR TITLE
Update tutorial programs to fix all broken dox links

### LINKED ISSUE #9460: [Step-20] Doxygen linking mechanism is inadequate when linking to different sections having the same heading name?
In the table of contents for Step-20, there are two sections with the same heading name

>1. Introduction
    - Linear solvers and preconditioners
2. The commented program
    - Linear solvers and preconditioners

Having read up to the `MixedLaplaceProblem::assemble_system` section in the commented code, I took a break and wanted to resume from the `Linear solvers and preconditioners` in the commented code. However, clicking this link in the TOC makes the page in the browser fly to the corresponding section name in the Introduction section! 

Assuming this is a bug in Doxygen, what would be the workaround in this kind of scenarios to help the reader with the navigation?

### PR DESCRIPTION
**Pull request by GrahamBenHarper**
Well, this fixes all of them *except* step-45. :stuck_out_tongue: Feel free to suggest better names than the ones I chose.

This closes #9460. These were spotted due to the work in #9505.

@masterleinad, how should we rename the sections in step-45?

### CODE REVIEW

File: contrib/utilities/checkdoxygen.py [hunk 1]
```diff
@@ -49,6 +49,37 @@ def check_empty_lines_in_tables(lines):
 
     return
 
+# have more than one header with the same name in a tutorial?
+def check_multiple_defined_headers(lines):
```
Comment (bangerth) [inline contrib/utilities/checkdoxygen.py]: I think you didn't want this commit as part of your patch -- let's leave it for a separate PR.

File: examples/step-20/step-20.cc [hunk 2]
```diff
@@ -559,7 +559,7 @@ namespace Step20
   }
 
 
-  // @sect3{Linear solvers and preconditioners}
+  // @sect3{Linear solvers and preconditioners implementation}
```
Comment (bangerth) [inline examples/step-20/step-20.cc]:
```suggestion
  // @sect3{Implementation of linear solvers and preconditioners}
```

File: examples/step-28/step-28.cc [hunk 3]
```diff
@@ -392,7 +392,7 @@ namespace Step28
   class EnergyGroup
   {
   public:
-    // @sect5{Public member functions}
+    // @sect5{<code>EnergyGroup</code> Public member functions}
```
Comment (bangerth) [inline examples/step-28/step-28.cc]:
```suggestion
    // @sect5{<code>EnergyGroup</code> public member functions}
```

File: examples/step-28/step-28.cc [hunk 4]
```diff
@@ -458,7 +458,7 @@ namespace Step28
                      const double         refine_threshold,
                      const double         coarsen_threshold);
 
-    // @sect5{Public data members}
+    // @sect5{<code>EnergyGroup</code> Public data members}
```
Comment (bangerth) [inline examples/step-28/step-28.cc]:
```suggestion
    // @sect5{<code>EnergyGroup</code> public data members}
```

File: examples/step-28/step-28.cc [hunk 5]
```diff
@@ -471,7 +471,7 @@ namespace Step28
     Vector<double> solution_old;
 
 
-    // @sect5{Private data members}
+    // @sect5{<code>EnergyGroup</code> Private data members}
```
Comment (bangerth) [inline examples/step-28/step-28.cc]:
```suggestion
    // @sect5{<code>EnergyGroup</code> private data members}
```

File: examples/step-28/step-28.cc [hunk 6]
```diff
@@ -501,7 +501,7 @@ namespace Step28
     AffineConstraints<double>                 hanging_node_constraints;
 
 
-    // @sect5{Private member functions}
+    // @sect5{<code>EnergyGroup</code> Private member functions}
```
Comment (bangerth) [inline examples/step-28/step-28.cc]:
```suggestion
    // @sect5{<code>EnergyGroup</code> private member functions}
```

File: examples/step-28/step-28.cc [hunk 7]
```diff
@@ -1170,7 +1170,7 @@ namespace Step28
     void run();
 
   private:
-    // @sect5{Private member functions}
+    // @sect5{<code>NeutronDiffusionProblem</code> Private member functions}
```
Comment (bangerth) [inline examples/step-28/step-28.cc]:
```suggestion
    // @sect5{<code>NeutronDiffusionProblem</code> private member functions}
```

File: examples/step-28/step-28.cc [hunk 8]
```diff
@@ -1183,7 +1183,7 @@ namespace Step28
     double get_total_fission_source() const;
 
 
-    // @sect5{Private member variables}
+    // @sect5{<code>NeutronDiffusionProblem </code> Private member variables}
```
Comment (bangerth) [inline examples/step-28/step-28.cc]:
```suggestion
    // @sect5{<code>NeutronDiffusionProblem </code> private member variables}
```

File: contrib/utilities/checkdoxygen.py (hunk 1 above)
Comment (GrahamBenHarper) [reply contrib/utilities/checkdoxygen.py]: Whoops! I did something funny with my branches and rebasing.

I created this branch at the tip of the script patch branch so I could spot which files had issues but rebasing on master still hangs onto those two commits from the script patch.

Comment (GrahamBenHarper) [reply contrib/utilities/checkdoxygen.py]: There we go :wink:

**Comment by GrahamBenHarper** [comment]
@bangerth I've made the requested changes
